### ISSUE TITLE
Load vectors disappeared - what is the alternative

### ISSUE THREAD

**Issue by kootenpv** [explosion/spaCy#1533]
As mentioned, in v2 these have disappeared:

```
load_vectors_from_bin_loc
load_vectors
```

What is the new way to enable loading different embeddings? I was using `googlenews.bin`.

My code was the following in `spacy_api`:

```python
def get_nlp(model="en", embeddings_path=None):
    import spacy
    if embeddings_path not in nlp_objects:
        if embeddings_path is None:
            nlp_ = spacy.load(model)
        else:
            if embeddings_path.endswith(".bin"):
                nlp_ = spacy.load(model, vectors=False)
                nlp_.vocab.load_vectors_from_bin_loc(embeddings_path)
            else:
                nlp_ = spacy.load(model, vectors=embeddings_path)
        nlp_objects[embeddings_path] = nlp_
    return nlp_objects[embeddings_path]
```

Please tell me it is still possible :)

**Comment by kootenpv** [comment]
I have tried to do a combination of using 1.9 vocab.save_dump and 2.0 from_bytes/from_disk, but I couldn't get it working and now I'm out of ideas.

**Comment by honnibal** [comment]
`vocab.vectors.data` is now just a numpy array, and `vocab.vectors.key2row` is a dict mapping hash values to row indices. So, any way you can get the data into that format will work. In v1 I used a pretty bad storage format where the keys were interleaved with the vectors data, which required pretty arbitrary C code to deserialize. If necessary I can bring back the reader for that format, but I'm hoping everyone can just provide their vectors in a more convenient format.

What format is `googlenews.bin`? What produced it?

**Comment by kootenpv** [comment]
The original is from here: https://github.com/mmihaltz/word2vec-GoogleNews-vectors

I followed this: https://github.com/explosion/spaCy/issues/884

Now I tried to create the `Vectors` in the new format by using spacy 1.9 to first save me the array, shape and keys. Since it is so large, it was my first usage of np.memmap :)

From the documentation I understood that keys was a `list` (quoting docs "A sequence of keys aligned with the data.").

I've created and dumped the vectors (didn't complain about list).

Then I did

```python
vectors = Vectors()
vectors = vectors.from_disk("/home/pascal/embeddings/googlenews_spacy")
nlp_ = spacy.load("en_core_web_sm", vectors=vectors)
```

`nlp_.vocab["hi"].vector` yields the error

```python
ValueError: Word vectors set to length 0. This may be because you don't have a model installed or loaded, or because your model doesn't include word vectors. For more info, see the documentation: 
https://spacy.io/usage/models
```

What could be the problem here? Is it the model that does not like it?
In case I should have indeed done a dict with string to int, then it should throw an error on a list :)
I'll generate a dict now.

**Comment by kootenpv** [comment]
Sometimes it can be easy.... unless you'll tell me it is wrong :)

I'm surprised though the above code doesn't work.

```python
nlp_ = spacy.load("en_core_web_sm", vectors=False)
nlp_.vocab.vectors = vectors
```

**Comment by honnibal** [comment]
@kootenpv The vectors API is surprisingly difficult to get right, because we need to balance not copying data against the usability. The solution we've ended up with is to avoid storing any strings in the vectors table, and avoid having the vectors own a `StringStore`. The vectors table itself is a relatively "dumb" store that just knows how to map the hashed keys to the rows of this blob of data.

This means that the vectors table is basically a child of the `Vocab`, so we don't want to have it passed in individually, or Bad Things might happen if the vectors has keys that the vocab doesn't know about. But if you want to write to the `nlp.vocab.vectors` attribute, well, that's okay. We follow the Python philosophy and assume you know what you're doing :p.

So in short: you can create and pass in a whole vocab, or you can replace the vectors table after the fact. Either should work fine.

**Comment by shengc** [comment]
@honnibal

I think this is actually broken in v2... 

Loading custom word vectors in v1 is possible:
https://stackoverflow.com/questions/42094180/spacy-how-to-load-google-news-word2vec-vectors

I am not sure if it is still possible in v2. If not, it is a very unfortunate regression ...

```python
In [2]: import gensim

In [3]: gensim.models.KeyedVectors.load_word2vec_format("GoogleNews-vectors-negative300.bin.gz", binary=True)
Out[3]: <gensim.models.keyedvectors.Word2VecKeyedVectors at 0x7f47beabcc18>

In [3]: w2v = Out[3]

In [5]: w2v.vectors.shape
Out[4]: (3000000, 300)

In [6]: import spacy

In [7]: nlp = spacy.load('en_core_web_md', vectors=False)

In [8]: nlp.vocab.vectors.shape
Out[5]: (20000, 300)

In [9]: nlp.vocab.vectors = w2v.vectors

In [10]: nlp.vocab['apple']
Out[6]: <spacy.lexeme.Lexeme at 0x7f469641e798>

In [11]: nlp.vocab['apple'].vector
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-16-e4de780e659d> in <module>()
----> 1 nlp.vocab['apple'].vector

lexeme.pyx in spacy.lexeme.Lexeme.vector.__get__()

vocab.pyx in spacy.vocab.Vocab.get_vector()

AttributeError: 'numpy.ndarray' object has no attribute 'key2row'
```

**Comment by nateraw** [comment]
Hey @shengc I was going through similar issues as you, and I think I have a solution for you! :smile: 

I took this idea from @honnibal here: 
#2032 

```python
import gensim
import spacy

# Path to google news vectors
google_news_path = "path\to\google\news\\GoogleNews-vectors-negative300.bin.gz"

# Load google news vecs in gensim
model = gensim.models.KeyedVectors.load_word2vec_format(gn_path, binary=True)

# Init blank english spacy nlp object
nlp = spacy.blank('en')

# Loop through range of all indexes, get words associated with each index.
# The words in the keys list will correspond to the order of the google embed matrix
keys = []
for idx in range(3000000):
    keys.append(model.index2word[idx])

# Set the vectors for our nlp object to the google news vectors
nlp.vocab.vectors = spacy.vocab.Vectors(data=model.syn0, keys=keys)

>>> nlp.vocab.vectors.shape
(3000000, 300)
```
